File: C++/Solution57.cpp
#include<iostream>
#include<vector>
#include<algorithm>
#include<unordered_map>
#include<unordered_set>
using namespace std;

class Solution57 {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        vector<vector<int>> res;
        int left = newInterval[0], right = newInterval[1];
        bool put = false;
        for(int index = 0; index < intervals.size(); index++){
            if((intervals[index][0] >= left && intervals[index][0] <= right) || (left >= intervals[index][0] && left <= intervals[index][1])){
                left = min(left, intervals[index][0]);
                right = max(right, intervals[index][1]);
                if(index == 3){
                cout << left << " " << right << endl;
            }
            }
            else if(left > intervals[index][0]){
                res.push_back(intervals[index]);
            }
            else{
                if(!put){
                    res.push_back({left, right});
                    put = true;
                }
                res.push_back(intervals[index]);
            }
        }
        if(!put){
            res.push_back({left, right});
        }
        return res;
    }
};

int main(){
    Solution57 s;
    vector<vector<int>> test = {{1, 2}, {3, 5}, {6, 7}, {8, 10}, {12, 16}};
    vector<int> interval = {4, 8};
    vector<vector<int>> res = s.insert(test, interval);
    return 0;
}

File: Java/other/Solution57.java
class Solution57 {
    public int[][] insert(int[][] intervals, int[] newInterval) {
    	if(intervals.length == 0){
    		return new int[][]{newInterval};
    	}
    	int[][] result = new int[intervals.length + 1][2];
    	int index = 0;
    	int left = newInterval[0];
    	int right = newInterval[1];
    	boolean done = false;
    	

    	for(int i = 0; i < intervals.length; i++){
    		if(intervals[i][1] < left){
    			result[index] = intervals[i];
    			index += 1;
    		}
    		else if(intervals[i][0] > right){
    			if(!done){
    				result[index] = new int[]{left, right};
    				index += 1;
    				done = true;
    			}
    			result[index] = intervals[i];
    			index += 1;
    		}
    		else{
    			
    			left = Math.min(left, intervals[i][0]);
    			right = Math.max(right, intervals[i][1]);
    		}
    	}
    	if(!done){
    		result[index] = new int[]{left, right};
    		index += 1;
    	}
    	int[][] copy = new int[index][2];
    	System.arraycopy(result, 0, copy, 0, index);
    	return copy;
    }

    public static void main(String[] args){
    	Solution57 s = new Solution57();
    	int[][] test = {{1, 5}};
    	int[] intervals = {6, 8};
    	int[][] result = s.insert(test, intervals);
    }
}